Client/game_sa/CFxSA.cpp

```diff
@@ -239,3 +246,37 @@ void CFxSA::TriggerFootSplash(CVector& vecPosition)
         call    dwFunc
     }
 }
+
+bool CFxSA::IsShadowsLimitReached()
+{
+    unsigned short& CShadows_ShadowsStoredToBeRendered = *(unsigned short*)0xC403DC;
+    return CShadows_ShadowsStoredToBeRendered >= 48;
+}
+
+bool CFxSA::AddShadow(eShadowType shadowType, CVector& vecPosition, CVector2D& vecOffset1, CVector2D& vecOffset2, SColor color, float fZDistance,
+                      bool bDrawOnWater, bool bDrawOnBuildings)
+{
+    if (IsShadowsLimitReached())
+        return false;
+
+
+    void*      textureAddress = *(void**)(SHADOW_BASE_TEXTURE_OFFSET + (int)shadowType * 4);
```

Comment:
It's better to use an array of RwTexture pointers here. Something like `g_ShadowTextures[shadowType]`.